src/lib/optimizer/strategy/dependent_group_by_reduction_rule.cpp

```diff
@@ -24,155 +24,103 @@ using namespace opossum;  // NOLINT
  *
  * @returns   Boolean value denoting whether at the group-by list of @param aggregate_node changed.
  */
-bool reduce_group_by_columns_for_constraint(const TableConstraintDefinition& table_constraint,
-                                            const std::set<ColumnID>& group_by_columns,
-                                            const std::shared_ptr<const StoredTableNode>& stored_table_node,
-                                            AggregateNode& aggregate_node) {
+bool reduce_group_by_columns_for_fd(const FunctionalDependency& fd, const ExpressionUnorderedSet& group_by_columns,
+                                    AggregateNode& aggregate_node) {
   auto group_by_list_changed = false;
-  const auto& constraint_columns = table_constraint.columns;
 
-  // Intersect primary key/unique columns and group-by columns. In case a primary key/unique constraint covers
-  // multiple columns, we need to check that all columns are present in order to later remove dependent columns.
-  std::vector<ColumnID> intersection;
-  std::set_intersection(constraint_columns.begin(), constraint_columns.end(), group_by_columns.begin(),
-                        group_by_columns.end(), std::back_inserter(intersection));
-
-  // Skip the current constraint as the primary key/unique constraint is not completely present.
-  if (intersection.size() != constraint_columns.size()) {
+  // To benefit from this rule, the FD's columns have to be part of the group-by list
+  if (!std::all_of(fd.first.cbegin(), fd.first.cend(),
+                   [&group_by_columns](std::shared_ptr<AbstractExpression> constraint_col_expr) {
+                     return group_by_columns.contains(constraint_col_expr);
+                   })) {
```

Comment:
If we make FunctionalDependencies a struct (to allow for `operator==`), this check should be done over there.

Reply:
never mind